### ISSUE TITLE
Level Crossing Links (Publisher/Subscriber) cannot be used in opaque composites

### ISSUE THREAD

**Issue by cxbrooks** [cxbrooks/test#97]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#165
From: @cxbrooks
Reported version: 7.1.devel
CC:   pt-dev@chess.eecs.berkeley.edu

Depends on:   [BZ#217](https://github.com/cxbrooks/test/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%23217%22)

**Comment by cxbrooks** [comment]
The problem is that if we have a large model and want to partition it
into many opaque composites that contain their own directors (say to
promote parallelism), then if the composites contain level crossing links
(aka liberal links) such as Publisher/Subscriber actors, we run into trouble.

Edward wrote:
> The mechanism we have cannot be made to work predictably across directors...
> I wouldn't know how to make this determinate.
>
> Would it work to add output ports at the boundary between directors,
> with a Subscriber on the inside and Publisher on the outside?

> I guess an alternative would be to change the mechanism to do the
> above automatically... This would be a fair amount of work, and fairly 
> tricky...

Brian Hudson wrote:
>    I forgot that Pub/Sub only works to the Director level, so if I have each
>    region using there own director, other regions can't see their publishers.
>
>    As stated in the Publisher docs:
>    *
>    "This actor publishes input tokens on a named channel. The tokens are
>    "tunneled" to any instance of Subscriber that names the same channel and
>    that is under the control of the same director. That is, it can be at a
>    different level of the hierarchy, or in an entirely different composite
>    actor, as long as the relevant composite actors are transparent (have no
>    director)."
>

**Comment by cxbrooks** [comment]
Making this bug depend on bug# 217 because if we can exportMoML on a model
with level crossing links, then the solution to this bug might be easier.

**Comment by cxbrooks** [comment]
*** Bug [BZ#341](https://github.com/cxbrooks/test/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%23341%22) has been marked as a duplicate of this bug. ***

**Comment by cxbrooks** [comment]
If we have a Publisher inside an TypedCompositeActor that has a Director
(making it opaque), then a Subscriber cannot Subscribe to the Publisher.

See
$PTII/ptolemy/actor/lib/test/auto/PublisherSubscriberOpaque.xml


Note that this sort of model sort of works in Modular CodeGen, but not in
regular interpreted mode.

**Comment by cxbrooks** [comment]
Pub/Sub can now be used in opaque components in interpreted mode.
See
$PTII/ptolemy/actor/lib/test/auto/PublisherSubscriberOpaque.xml
Dai did the bulk of the work here.

**Comment by cxbrooks** [comment]
Failed: 1  Total Tests: 225  ((Passed: 223, Newly Passed: 0)  Known Failed: 1) /
ome/bldmastr/ptII/ptolemy/actor/lib/test

==== Publisher-4.3 Channel Name Change: change the Publisher name in an Opaque model

    # Run the model from 4.2
    set manager [java::new ptolemy.actor.Manager [$model workspace] "p4manager"]
    $model setManager $manager
    $manager execute

==== Test generated error:

The width of relation .PublisherSubscriberOpaque.CompositeActor.publisherRelation2 can not be uniquely inferred.
Please make the width inference deterministic by explicitly specifying the width of this relation.The relation is deeply connected to these ports:
ptolemy.actor.TypedIOPort {.PublisherSubscriberOpaque.CompositeActor.Subscriber.input}
  in .PublisherSubscriberOpaque.CompositeActor.publisherRelation2


The problem here is that changing the channel name of an Publisher
in an Opaque requires that the model be written out and reloaded.
To replicate this:
$PTII/bin/vergil $PTII/ptolemy/actor/lib/test/auto/PublisherSubscriberOpaque.xml
1. Find the Publisher, change the channel name.
2. Find the Subscribe, change the channel name to the same name.
3. Run the model, you will get an error.

This bug is separate from my recent changes involving.
Publisher.propagateNameChanges

**Comment by cxbrooks** [comment]
The bug surround changing port names was fixed by Dai.

Another bug is that we seem to be generating code each time the model
is run

1.  $PTII/bin/vergil $PTII/ptolemy/cg/lib/test/auto/ModularCodeGen.xml
2.  Run the model, note that code is generated
3.  Exit
4.  $PTII/bin/vergil $PTII/ptolemy/cg/lib/test/auto/ModularCodeGen.xml
5.  Rerun the model,  note that code is generated again

In step 5, code should not be regenerated.

**Comment by cxbrooks** [comment]
I'm marking this as closed.  The bulk of the work was done by Dai.

More testing will uncover more bugs here.
